src/document/doc-utils.js

```diff
@@ -193,8 +193,26 @@ function getInnerParts(doc) {
   return parts;
 }
 
-function stripTrailingHardline(doc, withInnerParts = false) {
-  // HACK remove ending hardline, original PR: #1984
+// Remove empty docs
+function cleanDoc(doc) {
+  return mapDoc(doc, (currentDoc) => {
+    const { parts } = currentDoc;
+    if (Array.isArray(parts)) {
```

Comment:
`fill` has `parts` too and [you can't](https://github.com/prettier/prettier/blob/099c5fb7302021c894e38f53bb09909d45870502/src/document/doc-printer.js#L366-L367) remove individual parts from it that easy.

Reply:
Sorry, I didn't think though, empty `parts`  and empty strings in `concat` are safe to remove, right?

Reply:
Right. Turns out the existing function `normalizeDoc` has this bug too. Why not use this function in this PR, BTW?

Reply:
#9503 may be caused by this bug in `normalizeDoc` (or may be not).

Reply:
> Right. Turns out the existing function normalizeDoc has this bug too. Why not use this function in this PR, BTW?

I tried it first, not working(forgot how)